test/integration/test_display_callback.py

```diff
@@ -311,3 +317,95 @@ def test_module_level_no_log(executor, playbook):
     assert 'SENSITIVE' not in json.dumps(list(executor.events))
+
+
+def test_output_when_given_invalid_playbook(private_data_dir):
+    # As shown in the following issue:
+    #
+    #   https://github.com/ansible/ansible-runner/issues/29
+    #
+    # There was a lack of output by runner when a playbook that doesn't exist
+    # is provided.  This was fixed in this PR:
+    #
+    #   https://github.com/ansible/ansible-runner/pull/34
+    #
+    # But no test validated it.  This does that.
+    executor = init_runner(
+        private_data_dir=private_data_dir,
+        inventory="localhost ansible_connection=local",
+        envvars={"ANSIBLE_DEPRECATION_WARNINGS": "False"},
+        playbook=os.path.join(private_data_dir, 'fake_playbook.yml')
+    )
+
+    executor.run()
+    stdout = executor.stdout.read()
+    assert "ERROR! the playbook:" in stdout
+    assert "could not be found" in stdout
+
+
+def test_output_when_given_non_playbook_script(private_data_dir):
+    # As shown in the following pull request:
+    #
+    #   https://github.com/ansible/ansible-runner/pull/256
+    #
+    # This ports some functionality that previously lived in awx and allows raw
+    # lines of stdout to be treated as event lines.
+    #
+    # As mentioned in the pull request as well, there were no specs added, and
+    # this is a retro-active test based on the sample repo provided in the PR:
+    #
+    #   https://github.com/AlanCoding/ansible-runner-examples/tree/master/non_playbook/sleep_with_writes
+
+    with open(os.path.join(private_data_dir, "args"), 'w') as args_file:
+        args_file.write("bash sleep_and_write.sh\n")
+    with open(os.path.join(private_data_dir, "sleep_and_write.sh"), 'w') as script_file:
+        script_file.write("echo 'hi world'\nsleep 0.5\necho 'goodbye world'\n")
+
+    # Update the settings to make this test a bit faster :)
+    os.mkdir(os.path.join(private_data_dir, "env"))
+    with open(os.path.join(private_data_dir, "env", "settings"), 'w') as settings_file:
+        settings_file.write("pexpect_timeout: 0.2")
+
+    executor = init_runner(
+        private_data_dir=private_data_dir,
+        inventory="localhost ansible_connection=local",
+        envvars={"ANSIBLE_DEPRECATION_WARNINGS": "False"}
+    )
+
+    executor.run()
+    stdout = executor.stdout.readlines()
+    assert stdout[0].strip() == "hi world"
+    assert stdout[1].strip() == "goodbye world"
+
+    events = list(executor.events)
+
+    assert len(events) == 2
+    assert events[0]['event'] == 'verbose'
+    assert events[0]['stdout'] == 'hi world'
+    assert events[1]['event'] == 'verbose'
+    assert events[1]['stdout'] == 'goodbye world'
+
+
+@pytest.mark.parametrize('json_mode', [True])
+@pytest.mark.parametrize('playbook', [
+{'listvars.yml': '''
+- name: List Variables
+  connection: local
+  hosts: all
+  tasks:
+    - name: Display all variables/facts known for a host
+      debug:
+        var: hostvars[inventory_hostname]
```

Comment:
My preference for test writing here would be that we don't gather facts, and we don't rely on printing something from Ansible itself, which can change. It's possible to construct a string in Ansible itself of the desired length.

I ran this test and confirmed it failed, the JSON data got printed up to 4976 lines, totaling 18.8 k characters. I don't know what the interaction there is with pexpect's max 2k characters.

Reply:
Commenting in response to the above and https://github.com/ansible/ansible-runner/pull/335#issuecomment-698426758 as it seems to be about this test/comment as well.

* * *

> My preference for test writing here would be that we don't gather facts...

I think that is a fair point, and have no problem with going that route.  This was just a real world use case that I could point to in the tests to say how this could be triggered without needing any external dependencies to do so, but I also agree that relying on OS specific facts is a recipe for disaster, so I think using what you have in https://github.com/ansible/ansible-runner/pull/335#issuecomment-698426758 makes sense.

> I ran this test and confirmed it failed, the JSON data got printed up to 4976 lines, totaling 18.8 k characters. I don't know what the interaction there is with pexpect's max 2k characters.

I vaguely recall that happening for me as well and it was odd, but I am pretty sure there was a reason for it, but I don't recall what it was off hand (if I ever did figure it out).  There was a lot of time digging through the guts of `pexpect` that I didn't take notes on, so I can't remember off hand.

I think what ends up happening is it has something to do with this block of code:

https://github.com/ansible/ansible-runner/blob/f6f6a24538817a84d09feeac85a747b5b9238c0e/ansible_runner/utils.py#L256-L264

Where the `else` is triggered, and the result is that it prints both the line eventually, but also a duplicate of the JSON blob with the newlines not escaped (hence all the extra lines).  Again, this is just what I can vaguely remember, as it has been a while since I have looked at this issue.

* * *

My only suggestion with the suggested change of `"{{ ('F' * 150) | list }}"` would be to bump the 150 up to something like 4000 so we are for sure exceeding the 2k limit of `pexpect`'s default, but perhaps I am missing something with the `| list` portion of that line (which I admit to not fully understanding).